gflownet/envs/crystals/miller.py

```diff
@@ -0,0 +1,40 @@
+"""
+Class to represent an environment to sample Miller indices (hkl).
+"""
+from gflownet.envs.grid import Grid
+
+
+class MillerIndices(Grid):
+    """
+    The Miller indices are represented by either 3 parameters (h, k, l) if the
+    structure is cubic, or by 4 parameters (h, k, i, l) if the structure is hexagonal
+    or rhombohedral. Each parameter can take values in the set {-2, -1, 0, 1, 2}.
+    Therefore, we can represent the Miller indices environment by a hyper cube of
+    length 5, with dimensionality 3 or 4 depending on the structure.
+
+    Attributes
+    ----------
+    is_cubic : bool
+        True if the structure is cubic, False if the structure is hexagonal or
+        rhombohedral.
+
+    max_increment : int
+        Maximum increment of each dimension by the actions.
+
+    max_dim_per_action : int
+        Maximum number of dimensions to increment per action. If -1, then
+        max_dim_per_action is set to n_dim.
+    """
+
+    def __init__(
+        self,
+        is_cubic: bool,
+        max_increment: int = 1,
+        max_dim_per_action: int = 1,
+        **kwargs,
+    ):
+        if is_cubic:
+            n_dim = 3
+        else:
+            n_dim = 4
```

Comment:
I would suggest to actually always use `n_dim=3`. My reasoning : 

In the hexagonal/rhombohedral scenario, there are actually a constraint on the values of the Miller indices since there is some redundancy between them. The constraint is `h + k + i = 0.` 
Source : https://dictionary.iucr.org/Miller_indices#Bravais-Miller_indices_.28hexagonal_axes.29

That means that there are only ever 3 dimensions that we need to pick : 
- `h, k, i` for non-{hexagonal, rhombohedral}
- `h, k, l` for {hexagonal, rhombohedral}

In the hexagonal/rhombohedral scenario, we can simply infer the value of `i` from the values of `h` and `k`. The only thing that we need to do is ensure that `h` and `k` don't take values that makes it impossible to find a valid value of `i`. For that, I think we only need to ensure that `-2 <= h+k <=2`.

Reply:
If we do this change, we might need to add tests to make sure that we don't allow the picking of, say, `h=2 and k=2` for a rhombohedral structure.

Reply:
You are right, I had overlooked all this :( Then we need to make several changes here.